### PR TITLE
gh-124190: Ignore files directories check warning tooling

### PR DESCRIPTION
**Pull request by nohlson**
Add the ability to ignore entire files regardless of the warning count and entire directories in the check compiler warning CI tooling. This will facilitate ignoring warnings generated from vendored code, or test/test generation related code.

* Issue: gh-124190

### CODE REVIEW

File: Tools/build/check_warnings.py [hunk 1]
```diff
@@ -11,9 +11,49 @@
...
+class IgnoreRule(NamedTuple):
+    file_path: str
     count: int
+    ignore_all: bool = False
+    is_directory: bool = False
+
+
+def parse_warning_ignore_file(file_path: str) -> set[IgnoreRule]:
+    """
+    Parses the warning ignore file and returns a set of IgnoreRules
```
Comment (rruuaanng) [inline Tools/build/check_warnings.py:23]: Your document string is missing a description of the argument.
emmm, I think that's right.

File: Tools/build/check_warnings.py [hunk 2]
```diff
@@ -78,9 +122,24 @@ def get_warnings_by_file(warnings: list[dict]) -> dict[str, list[dict]]:
...
+) -> IgnoreRule | None:
+    """
+    Returns the IgnoreRule object for the file path if there is a related rule for it
+    """
+    for ignore_rule in ignore_rules:
+        if ignore_rule.is_directory:
+            if file_path.startswith(ignore_rule.file_path):
+                return ignore_rule
+        elif file_path == ignore_rule.file_path:
+            return ignore_rule
```
Comment (hugovk) [inline Tools/build/check_warnings.py]: Let's follow an example below, just call it `rule`, we only have one kind of rule and makes it a bit easier to read:
```suggestion
    for rule in ignore_rules:
        if rule.is_directory:
            if file_path.startswith(rule.file_path):
                return rule
        elif file_path == rule.file_path:
            return rule
```

File: Tools/build/check_warnings.py [hunk 3]
```diff
@@ -89,14 +148,21 @@ def get_unexpected_warnings(
...
+            if ignore_rule.ignore_all:
+                continue
+
+            if len(files_with_warnings[file]) > ignore_rule.count:
+                unexpected_warnings[file] = (
+                    files_with_warnings[file],
+                    ignore_rule.count,
+                )
+            continue
+        elif ignore_rule is None:
```
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
        rule = is_file_ignored(file, ignore_rules)

        if rule:
            if rule.ignore_all:
                continue

            if len(files_with_warnings[file]) > rule.count:
                unexpected_warnings[file] = (
                    files_with_warnings[file],
                    rule.count,
                )
            continue
        elif rule is None:
```

File: Tools/build/check_warnings.py (hunk 2 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: @hugovk I agree. renamed to `rule`

Comment (nohlson) [reply Tools/build/check_warnings.py]: or did you mean every instance of `ignore_rule` change to `rule`

File: Tools/build/check_warnings.py (hunk 3 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: Renamed `ignore_rule` to `rule`

File: Tools/build/check_warnings.py (hunk 2 above)
Comment (nohlson) [reply Tools/build/check_warnings.py]: 👍 understood

**Review by hugovk — APPROVED** [review]
Thanks!

**Pull request merged**